Ck tile gemm cshuffle & CK Tile GEMM restructure

ThomasNing opened the pull request:
There are several changes of this pull request:
1. Delete the Layouts in the kernel input template
2. Add feature that CK Tile GEMM CPU and GPU reference to support new layout
3. Rename the CK Tile GEMM pipeline method
4. Finished the first step of the c shuffle epilogue that we could do the permutation in either the tile level of all the tiles in the kernel or the data level inside one tile

ToDo on next PR:
1. C Shuffle Performance improvement on regroup the tile data in the shared memory

ThomasNing committed:
ake the cshuffle compilable

ThomasNing committed:
modify Mhe reference on gpu and cpu. Correaccess of cshuffle

ThomasNing committed:
fix the cpu reference code

ThomasNing committed:
Complete the in tile shuffle logic

ThomasNing committed:
restructure the kernel template input

ThomasNing committed:
change the naming pattern of ck_tile gemm pipeline

ThomasNing commented:
@carlushuang @aosewski  PTAL Thanks!

poyenc committed:
Re-format files using remod.py

aosewski reviewed (commented):
Hi Thomas, I'm not sure if this is what cshuffle should actually do. I don't think you should permute data by hand actually and especially permute data hold by just single thread. The CShuffle should collectively (by all threads in workgroup) first store data to LDS and then again collectively read the data from LDS (with different distribution) and write data to global memory.

ThomasNing commented:
@aosewski Yes I totally agree with you adam. Since we had the discussion last week, I have addressed your comments and talked with Carlus. We decided to separate the feature to two parts. One is get the data with new permutation from the tile level. The next one is reorganized that in the shared memory. This merge request is for the first phase. I will have the next one after this one is merged. Because this one has the other features need to merge first. The shared memory reorganization will add next in the cshuffle epilogue.

ThomasNing committed:
Solve the fmha conflict with gemm

ThomasNing committed:
Merge branch 'ck_tile_gemm_cshuffle' of https://github.com/ROCm/composable_kernel into ck_tile_gemm_cshuffle

ThomasNing committed:
Merge branch 'develop' into ck_tile_gemm_cshuffle

carlushuang commented on include/ck_tile/core/container/thread_buffer.hpp:
```diff
@@ -59,6 +59,14 @@ struct thread_buffer {
     template <index_t I> CK_TILE_HOST_DEVICE constexpr auto& at(number<I>)              { return get(I); }
     template <index_t I> CK_TILE_HOST_DEVICE constexpr const auto& at(number<I>) const  { return get(I); }
 
+    CK_TILE_HOST_DEVICE constexpr void set_value(index_t i, const value_type& value) { 
```
we have `set_as()` API in below, do we need a new one here? I feel you can just reuse that in below

ThomasNing replied:
Fixed

carlushuang reviewed (commented):
![image](https://github.com/user-attachments/assets/3ffd0a4d-b563-49cd-b7fb-d626d4fa549e)
I think you need fix all the `BlockGemmPipeline* -> GemmPipeline*` in this PR, otherwise CI will not pass

ThomasNing committed:
Comment Addressed from Carlus

carlushuang commented:
cc @danyao12

ThomasNing committed:
Merge branch 'develop' into ck_tile_gemm_cshuffle

ThomasNing commented:
@danyao12 @qianfengz The CI has passed please take a look.

danyao12 commented on include/ck_tile/ops/fmha/pipeline/block_fmha_bwd_pipeline_default_policy.hpp:
```diff
@@ -52,21 +59,27 @@ struct BlockFmhaBwdPipelineDefaultPolicy
+                                                       Problem::BlockFmhaShape::kK1>,
+                                              typename Problem::BlockFmhaShape::Gemm1BlockWarps,
+                                              typename Problem::BlockFmhaShape::Gemm1WarpTile>,
+                                TileGemmTraits<Problem::kPadSeqLenQ,
```
These TileGemmTraits are useless in fmha fwd&bwd policy, and the kPadX_ & LayoutX_ may not match the real Gemm implementation in fmha and can be confusing. Maybe you can set kPadX_ to false and LayoutX_ to RowMajor with comment that these settings are useless. @qianfengz @poyenc What do you think?

ThomasNing replied:
Thank you for the comment Dan! I just discussed with Qianfeng offline. I will call the Block level BlockGemmProblem to avoid the useless information. The fmha only cares about block level gemm so I will let it call the block-level gemm  problem. All the previous version call it on the pipeline level. Thank you for the comment!

carlushuang reviewed (approved):
LGTM
